tests/integration/cli/commands/test_issue_2519.py

```diff
@@ -0,0 +1,157 @@
+        "bloom-filter2>=2.0.0",
+        "--indent",
+        "2",
+        "--lock",
+        lock,
+        *extra_args
+    )
+    result.assert_success()
+    return result
+
+
+TARGET = LocalInterpreter.create()
+PLATFORM_TAG = TARGET.platform.tag
+
+skip_if_py2 = pytest.mark.skipif(
+    PY2, reason="The bloom-filter2 project only claims support for Python 3.x"
+)
+
+
+@skip_if_py2
+@pytest.mark.parametrize(
+    "pip_version",
+    [
+        pytest.param(version, id=str(version))
+        for version in PipVersion.values()
+        if version < PipVersion.v23_2 and version.requires_python_applies(TARGET)
+    ],
+)
+def test_sync_version_mismatch_old_pip(
+    pip_version,  # type: PipVersionValue
+    tmpdir,  # type: Any
+):
+    # type: (...) -> None
+
+    lock = os.path.join(str(tmpdir), "lockfile.json")
+    assert_lock_sync(lock, pip_version)
+
+    result = assert_lock_sync(lock, pip_version)
+    assert re.match(
+        r"^{lead_in}\n"
+        r"    \+ http.*{wheel}\n"
+        r"    - http.*{sdist}$\n".format(
+            lead_in=re.escape(
+                "Updates for lock generated by {platform_tag}:\n"
+                "  Updated bloom-filter2 2 artifacts:".format(platform_tag=PLATFORM_TAG)
+            ),
+            wheel=re.escape("bloom_filter2-2.0.0-py3-none-any.whl"),
+            sdist=re.escape("bloom-filter2-2.0.0-1.tar.gz"),
+        ),
+        result.error,
+    ), result.error
+
+    result = assert_lock_sync(lock, pip_version)
+    assert (
+        "No updates for lock generated by {platform_tag}.\n".format(platform_tag=PLATFORM_TAG)
+        == result.error
+    )
+
+    result = assert_lock_sync(lock, pip_version)
+    assert (
+        "No updates for lock generated by {platform_tag}.\n".format(platform_tag=PLATFORM_TAG)
+        == result.error
+    )
+
+
+@skip_if_py2
+@pytest.mark.parametrize(
+    "pip_version",
+    [
+        pytest.param(version, id=str(version))
+        for version in PipVersion.values()
+        if version >= PipVersion.v23_2 and version.requires_python_applies(TARGET)
+    ],
+)
+def test_sync_version_mismatch_new_pip(
+    pip_version,  # type: PipVersionValue
+    tmpdir,  # type: Any
+):
+    # type: (...) -> None
+
+    lock = os.path.join(str(tmpdir), "lockfile.json")
+    assert_lock_sync(lock, pip_version)
+
+    result = assert_lock_sync(lock, pip_version)
+    assert (
+        "No updates for lock generated by {platform_tag}.\n".format(platform_tag=PLATFORM_TAG)
+        == result.error
+    )
+
+    result = assert_lock_sync(lock, pip_version)
+    assert (
+        "No updates for lock generated by {platform_tag}.\n".format(platform_tag=PLATFORM_TAG)
+        == result.error
+    )
+
+
+@skip_if_py2
+@pytest.mark.parametrize(
+    "pip_version",
+    [
+        pytest.param(version, id=str(version))
+        for version in PipVersion.values()
+        if version.requires_python_applies(TARGET)
+    ],
+)
+def test_sync_version_mismatch_prefer_binary_all_pips(
```

Comment:
This test demonstrates the alternate workaround to using a newer Pip described in #2519, using `--prefer-binary` instead. Note that this works over all Pip versions, but is not typically generic. You'd only want to apply it for certain resolves.